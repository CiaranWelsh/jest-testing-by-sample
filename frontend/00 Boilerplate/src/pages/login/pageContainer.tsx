import * as React from 'react';
import { history } from '../../history';
import {
  LoginEntity, createEmptyLoginEntity,
  LoginFormErrors, createEmptyLoginFormErrors,
} from './viewModel';
import { validations } from './validations';
import { LoginPage } from './page';
import { routes } from '../../common/constants/routes';

interface State {
  loginEntity: LoginEntity;
  loginFormErrors: LoginFormErrors;
}

export class LoginPageContainer extends React.PureComponent<{}, State> {
  state = {
    loginEntity: createEmptyLoginEntity(),
    loginFormErrors: createEmptyLoginFormErrors(),
  };

  updateField = (fieldName: string, value: any) => {
    validations.validateField(this.state.loginEntity, fieldName, value)
      .then((fieldValidationResult) => {
        this.setState({
          loginEntity: {
            ...this.state.loginEntity,
            [fieldName]: value,
          },
          loginFormErrors: {
            ...this.state.loginFormErrors,
            [fieldName]: fieldValidationResult,
          },
        });
      });
  }

  doLogin = () => {
    validations.validateForm(this.state.loginEntity)
      .then((formValidationResult) => {
        if (formValidationResult.succeeded) {
          history.push(routes.members);
        } else {
          alert('error, review the fields');
        }
      });
  }

  render() {
    return (
      <LoginPage
        loginEntity={this.state.loginEntity}
        loginFormErrors={this.state.loginFormErrors}
        updateField={this.updateField}
        doLogin={this.doLogin}
      />
    );
  }
}
