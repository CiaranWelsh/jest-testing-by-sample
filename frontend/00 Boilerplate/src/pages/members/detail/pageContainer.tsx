import * as React from 'react';
import { MemberDetailPage } from './page';

export class MemberDetailPageContainer extends React.PureComponent {
  render() {
    return (
      <MemberDetailPage
      />
    );
  }
}
