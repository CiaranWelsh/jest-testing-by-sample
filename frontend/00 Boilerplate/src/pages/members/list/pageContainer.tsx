import * as React from 'react';
import { MemberListPage } from './page';

export class MemberListPageContainer extends React.PureComponent {
  render() {
    return (
      <MemberListPage
      />
    );
  }
}
