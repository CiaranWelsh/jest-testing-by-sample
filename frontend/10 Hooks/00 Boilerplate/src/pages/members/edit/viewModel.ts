export interface Member {
  id: number;
  name: string;
  avatarUrl: string;
}

export const createEmptyMember = (): Member => ({
  id: -1,
  name: '',
  avatarUrl: '',
});
