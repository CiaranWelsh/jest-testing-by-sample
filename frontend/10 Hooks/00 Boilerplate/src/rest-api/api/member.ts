import { Member } from '../model/member';

const fetchMemberListUrl = 'https://api.github.com/orgs/lemoncode/members';
const fetchMemberUrl = 'https://api.github.com/users/';

export const fetchMembers = (): Promise<Member[]> =>
  fetch(fetchMemberListUrl).then(extractPayload);

export const fetchMember = (name: string): Promise<Member> =>
  fetch(`${fetchMemberUrl}${name}`).then(extractPayload);

const extractPayload = (response: Response): Promise<any> =>
  response.ok ? response.json() : responseError(response);

const responseError = (response: Response): Promise<any> =>
  response.json().then(error => Promise.reject(error.message));
