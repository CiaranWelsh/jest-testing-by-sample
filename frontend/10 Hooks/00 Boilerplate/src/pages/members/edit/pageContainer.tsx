import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { fetchMember } from '../../../rest-api/api/member';
import { history } from '../../../history';
import { mapMemberModelToVM } from './mappers';
import { createEmptyMember } from './viewModel';
import { MemberEditPage } from './page';

const useLoadMember = (name: string) => {
  const [member, setMember] = React.useState(createEmptyMember());

  React.useEffect(() => {
    fetchMember(name).then(modelMember => {
      setMember(mapMemberModelToVM(modelMember));
    });
  }, [name]);

  return { member, setMember };
};

interface Params {
  name: string;
}

export const MemberEditPageContainer: React.StatelessComponent<
  RouteComponentProps<Params>
> = props => {
  const { member, setMember } = useLoadMember(props.match.params.name);
  return (
    <MemberEditPage
      member={member}
      onUpdateField={(field, value) => {
        setMember({ ...member, [field]: value });
      }}
      onSave={() => {
        history.goBack();
      }}
    />
  );
};
