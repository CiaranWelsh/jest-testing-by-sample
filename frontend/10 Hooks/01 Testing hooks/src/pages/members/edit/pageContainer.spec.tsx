import * as React from 'react';
import { shallow, mount } from 'enzyme';
import * as memberAPI from '../../../rest-api/api/member';
import * as model from '../../../rest-api/model';
import * as vm from './viewModel';
import * as mappers from './mappers';
import { history } from '../../../history';
import { MemberEditPageContainer } from './pageContainer';

describe('pages/members/edit/pageContainer tests', () => {
  it('should render as expected feeding required properties', () => {
    // Arrange
    const props: any = {
      match: {
        params: {
          name: 'test user name',
        },
      },
    };

    // Act
    const component = shallow(<MemberEditPageContainer {...props} />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  it('should call to fetchMember on mount', () => {
    // Arrange
    const props: any = {
      match: {
        params: {
          name: 'test user name',
        },
      },
    };

    const mockMember: model.Member = {
      id: 1,
      login: 'testLogin',
      avatar_url: 'testAvatarUrl',
    };
    const fetchMemberStub = jest
      .spyOn(memberAPI, 'fetchMember')
      .mockResolvedValue(mockMember);

    // Act
    const component = mount(<MemberEditPageContainer {...props} />);

    // Assert
    expect(fetchMemberStub).toHaveBeenCalledWith(props.match.params.name);
  });

  it('should call to mapMemberModelToVM on mount', done => {
    // Arrange
    const props: any = {
      match: {
        params: {
          name: 'test user name',
        },
      },
    };

    const mockMember: model.Member = {
      id: 1,
      login: 'testLogin',
      avatar_url: 'testAvatarUrl',
    };
    const fetchMemberStub = jest
      .spyOn(memberAPI, 'fetchMember')
      .mockResolvedValue(mockMember);

    const mockVmMember: vm.Member = {
      id: 1,
      name: 'testName',
      avatarUrl: 'testAvatarUrl',
    };
    const mapperStub = jest
      .spyOn(mappers, 'mapMemberModelToVM')
      .mockReturnValue(mockVmMember);

    // Act
    const component = mount(<MemberEditPageContainer {...props} />);

    // Assert
    setImmediate(() => {
      expect(mapperStub).toHaveBeenCalledWith(mockMember);
      done();
    });
  });

  it('should update member when fire onUpdateField', () => {
    // Arrange
    const props: any = {
      match: {
        params: {
          name: 'test user name',
        },
      },
    };

    // Act
    const component = shallow(<MemberEditPageContainer {...props} />);

    component.prop('onUpdateField')('name', 'updated name');

    // Assert
    expect(component).toMatchSnapshot();
  });

  it('should call to history goBack when fire onSave', () => {
    // Arrange
    const props: any = {
      match: {
        params: {
          name: 'test user name',
        },
      },
    };

    const goBackStub = jest.spyOn(history, 'goBack')
    .mockImplementation(() => {});

    // Act
    const component = shallow(<MemberEditPageContainer {...props} />);

    component.prop('onSave')();

    // Assert
    expect(goBackStub).toHaveBeenCalled();
  });
});
