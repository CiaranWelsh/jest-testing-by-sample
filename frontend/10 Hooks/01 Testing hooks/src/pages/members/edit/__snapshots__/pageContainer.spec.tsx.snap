// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/members/edit/pageContainer tests should render as expected feeding required properties 1`] = `
<Component
  member={
    Object {
      "avatarUrl": "",
      "id": -1,
      "name": "",
    }
  }
  onSave={[Function]}
  onUpdateField={[Function]}
/>
`;

exports[`pages/members/edit/pageContainer tests should update member when fire onUpdateField 1`] = `
<Component
  member={
    Object {
      "avatarUrl": "",
      "id": -1,
      "name": "updated name",
    }
  }
  onSave={[Function]}
  onUpdateField={[Function]}
/>
`;
